---
title: "Seurat - Guided Clustering Tutorial"
author: "Michael Peters"
date: "3/17/2020"
output: html_document
---

## Setup the Seurat Object

We are analyzing a dataset of PBMCs from 10X Genomics. There are 2,700 single cells that were sequenced on the Illumina NextSeq 500.


### Start by reading in the data:

* The `Read10X` function reads in the output of the **cellranger** pipeline from 10X and returns a unique molecular identified (UMI) matrix.
    + The values of a **UMI count matrix** represent the number of molecules for each feature (i.e. gene; row) that are detected in each cell (column).
* We then use the count matrix to create a `Seurat` object.
    + This serves as a container that contains both data (like the count matrix) and anlaysis (like PCA or clustering results) for a single-cell dataset.
        - Example: `pbmc[["RNA"]]@counts` stores the count matrix
        
```{r Load Packages, message=FALSE, warning=FALSE}

library(dplyr)
library(Seurat)
library(patchwork)

```

```{r Setup}

# Check if folder exists and download if not.
if(!dir.exists("./data/pbmc3k/")){
    dir.create("./data/")
    dir.create("./data/pbmc3k/")
    }

if(!dir.exists("./output/")){
    dir.create("./output/")
    }

if(!file.exists("./data/pbmc3k/pbmc3k_filtered_gene_bc_matrices.tar.gz")) {
  download.file("https://s3-us-west-2.amazonaws.com/10x.files/samples/cell/pbmc3k/pbmc3k_filtered_gene_bc_matrices.tar.gz",
  destfile = "./data/pbmc3k/pbmc3k_filtered_gene_bc_matrices.tar.gz") }

untar("./data/pbmc3k/pbmc3k_filtered_gene_bc_matrices.tar.gz", exdir = "./data/pbmc3k/")

# Load the PBMC dataset
pbmc.data <- Read10X(data.dir = "./data/pbmc3k/filtered_gene_bc_matrices/hg19/")

# Initialize the Seurat object with the raw (non-normalized) data.
pbmc <- CreateSeuratObject(counts = pbmc.data, project = "pbmc3k", min.cells = 3, min.features = 200)
pbmc

```

### Previewing data in a count matrix

```{r Preview Data}

# Load a few genes in the first thirty cells
pbmc.data[c("CD3D", "TCL1A", "MS4A1"), 1:30]

```

The `.` values represent 0s (no molecules detected). Most values in an scRNA-seq matrix are 0, so Seurat uses a sparse-matrix representation whenever possible.

### Examine the size differences of this dataset using a dense matrix and a sparse matrix.

```{r Dense vs Sparse Sizes}
dense.size <- object.size(as.matrix(pbmc.data))
dense.size

sparse.size <- object.size(pbmc.data)
sparse.size

dense.size/sparse.size

```

## Standard Pre-Processing Workflow

These steps outline the standard pre-processing workflow for scRNA-seq data in Seurat. They represent the selection and filtration of cells based on:

* QC metrics
* Data normalization and scaling
* The detection of highly variable features.


### QC and selecting cells for further analysis

Seurat lets you to easily explore QC metrics and filter cells based on any user-defined criteria. A few QC metrics commonly used by the community include:

* The number of unique genes detected in each cell.
    + Low-quality cells or empty droplets will often have very few genes
    + Cell doublets or multiplets may exhibit an aberrantly high gene count
* Similarly, the total number of molecules detected within a cell (correlates strongly with unique genes)
* The percentage of reads that map to the mitochondrial genome
    + Low-quality / dying cells often exhibit extensive mitochondrial contamination
    + We calculate mitochondrial QC metrics with the `PercentageFeatureSet` function, which calculates the percentage of counts originating from a set of features
    + We use the set of all genes starting with `MT-` as a set of mitochondrial genes

```{r QC and Cell Selecting}

# The [[ operator can add columns to object metadata. This is a great place to stash QC stats
pbmc[["percent.mt"]] <- PercentageFeatureSet(pbmc, pattern = "^MT-")

```

**Where does Seurat store QC metrics?**

The number of unique genes and total molecules are automatically calculated during `CreateSeuratObject`
* They can be found stored in the object meta data

```{r Show Some QC Metrics}

# Show QC metrics for the first 5 cells
head(pbmc@meta.data, 5)

```

Example visualizing QC metrics and using these to filter cells:

* We filter cells that have unique feature counts over 2,500 or less than 200
* We filter cells that have >5% mitochondrial counts

```{r Visualize QC Metrics}

# Visualize QC metrics as a violin plot
VlnPlot(pbmc, features = c("nFeature_RNA", "nCount_RNA", "percent.mt"), ncol = 3)

```

```{r FeatureScatter}

# FeatureScatter is typically used to visualize feature-feature relationships, but can be used for anything calculated by the object, i.e. columns in object metadata, PC scores etc.

plot1 <- FeatureScatter(pbmc, feature1 = "nCount_RNA", feature2 = "percent.mt")
plot2 <- FeatureScatter(pbmc, feature1 = "nCount_RNA", feature2 = "nFeature_RNA")

plot1 + plot2
```

```{r Remove Unwanted Data}

pbmc <- subset(pbmc, subset = nFeature_RNA > 200 & nFeature_RNA < 2500 & percent.mt < 5)

```

### Normalizing the Data

After unwanted cells are removed from the dataset, the data needs to be normalized. A global-scaling normalization `LogNormalize` is used by default. This normalizes the feature expression measurements for each cell by the total expression, multiplies it by a scale factor (10,000 by default), and log-transforms the result. Normalized values are stored in `pbmc[["RNA"]]@data`

```{r NormalizeData, Showing Defaults}

pbmc <- NormalizeData(pbmc, normalization.method = "LogNormalize", scale.factor = 10000)

```

For clarity, in this previous line of code (and in future commands), we provide the default values for certain parameters in the function call. However, this isn’t required and the same behavior can be achieved with:

```{r NormalizeData, Hiding Defaults}

pbmc <- NormalizeData(pbmc)

```

### Identification of Highly Variable Features (Feature Selection)

We next calculate a subset of features that exhibit high cell-to-cell variation in the dataset (i.e, they are highly expressed in some cells, and lowly expressed in others).

* Focusing on these genes in downstream analysis helps to highlight biological signal in single-cell datasets.

The Seurat3 procedure modelings the mean-variance relationship inherent in single-cell data, and is implemented in the `FindVariableFeatures` function. By default, 2,000 features per dataset are returned. These will be used in downstream analysis, like PCA.

```{r Find and Plot Variable Features}

pbmc <- FindVariableFeatures(pbmc, selection.method = "vst", nfeatures = 2000)

# Identify the 10 most highly variable genes
top10 <- head(VariableFeatures(pbmc), 10)

# plot variable features with and without labels
plot1 <- VariableFeaturePlot(pbmc)
plot2 <- LabelPoints(plot = plot1, points = top10, repel = TRUE)
plot1 + plot2

```

### Scaling the Data

A standard pre-procesing step prior to dimensional reduction techinques like pca, linear transformation (`scaling`), is applied. The `scale data` function:

* Shifts the expression of each gene so that the mean expression across cells is 0
* Scales the expression of each gene so that the variance across cells is 1
    + This gives equal weight in downstream analysis so that highly-expressed genes do not dominate
* The results are stored in `pbmc[["RNA"]]@scale.data`

```{r Scale Data}

all.genes <- rownames(pbmc)
pbmc <- ScaleData(pbmc, features = all.genes)

```

### Perform Linear Dimensional Reduction

Now we perform PCA on the scaled data. By default, only the previously determined variable features are used as input, but can be defined using features argument if you wish to choose a different subset.

```{r Run a PCA}

pbmc <- RunPCA(pbmc, features = VariableFeatures(object = pbmc))

```

Seurat provides several useful ways of visualizing both cells and features that define the PCA, including `VizDimReduction`, `DimPlot`, and `DimHeatmap`

**Examine and visualize PCA results a few different ways**

```{r Print PCA Results}

print(pbmc[["pca"]], dims = 1:5, nfeatures = 5)

```

```{r  Visualize Dimension Loadings}

VizDimLoadings(pbmc, dims = 1:2, reduction = "pca")

```

```{r Dimension Plot}

DimPlot(pbmc, reduction = "pca")

```

`DimHeatmap` allows for easy exploration of the primary sources of heterogeneity in a dataset, and can be useful when trying to decide which PCs to include for further downstream analyses. 

* Both cells and features are ordered according to their PCA scores. Setting `cells` to a number plots the ‘extreme’ cells on both ends of the spectrum, which dramatically speeds plotting for large datasets. Though clearly a supervised analysis, this is a valuable tool for exploring correlated feature sets.

```{r Single-Dimension Heatmap}

DimHeatmap(pbmc, dims = 1, cells = 500, balanced = TRUE)

```


```{r Multiple-Dimension Heatmap}

DimHeatmap(pbmc, dims = 1:15, cells = 500, balanced = TRUE)

```

### Determine the 'Significant' of the Dataset

To overcome the extensive technical noise in any single feature for scRNA-seq data, Seurat clusters cells based on their PCA scores, with each PC essentially representing a ‘metafeature’ that combines information across a correlated feature set. The top principal components therefore represent a robust compression of the dataset. However, how many componenets should we choose to include? 10? 20? 100?

A resampling test is used to randomly permute a subset of the data (1% by default) and rerun PCA, constructing a ‘null distribution’ of feature scores, and repeat this procedure. **‘Significant’ PCs are identified as those who have a strong enrichment of low p-value features.**

```{r JackStraw Procedure}

# NOTE: This process can take a long time for big datasets, comment out for expediency. More approximate techniques such as those implemented in ElbowPlot() can be used to reduce computation time

pbmc <- JackStraw(pbmc, num.replicate = 100)
pbmc <- ScoreJackStraw(pbmc, dims = 1:20)

```

The `JackStrawPlot` function provides a visualization tool for comparing the distribution of p-values for each PC with a uniform distribution (dashed line). **‘Significant’ PCs will show a strong enrichment of features with low p-values (solid curve above the dashed line).** In this case it appears that there is a sharp drop-off in significance after the first 10-12 PCs.

```{r JackStrawPlot}

JackStrawPlot(pbmc, dims = 1:15)

```

An alternative heuristic method generates an ‘Elbow plot’: a ranking of principle components based on the percentage of variance explained by each one (`ElbowPlot` function). In this example, we can observe an ‘elbow’ around PC 9-10, suggesting that the majority of true signal is captured in the first 10 PCs.

```{r Elbow Plot}

ElbowPlot(pbmc)

```

Use these three approaches to determine the true dimensionality of a dataset.

* The first is more supervised, exploring PCs to determine relevant sources of heterogeneity, and could be used in conjunction with GSEA for example. 
* The second implements a statistical test based on a random null model, but is time-consuming for large datasets, and may not return a clear PC cutoff. 
* The third is a heuristic that is commonly used, and can be calculated instantly. In this example, all three approaches yielded similar results, but we might have been justified in choosing anything between PC 7-12 as a cutoff.

### Cluster the Cells

We first construct a KNN graph based on the the euclidean distance in PCA space, and refine the edge weights between any two cells based on the shared overlap in their local neighborhoods. This step is performed using the `FindNeighbors` function, and takes as input the previously defined dimensionality of the dataset (first 10 PCs).

To cluster the cells, we next apply modularity optimization techniques such as the Louvain algorithm (default) or SLM, to iteratively group cells together, with the goal of optimizing the standard modularity function. The `FindClusters` function implements this procedure, and contains a resolution parameter that sets the ‘granularity’ of the downstream clustering, with increased values leading to a greater number of clusters. We find that setting this parameter between 0.4-1.2 typically returns good results for single-cell datasets of around 3K cells. Optimal resolution often increases for larger datasets. The clusters can be found using the `Idents` function.

```{r Find Neighbors and Clusters}

pbmc <- FindNeighbors(pbmc, dims = 1:10)
pbmc <- FindClusters(pbmc, resolution = 0.5)

```

```{r Look at Some Cluster IDs}

# Look at cluster IDs of the first 5 cells
head(Idents(pbmc), 5)

```

### Run Non-Linear Dimensional Reduction (UMAP/tSNE)

UMAP and tSNE can be used to visualize and explore datasets. The goal of these algorithms is to learn the underlying manifold of the data in order to place similar cells together in low-dimensional space. Cells within the graph-based clusters determined above should co-localize on these dimension reduction plots. As input to the UMAP and tSNE, we suggest using the same PCs as input to the clustering analysis.

```{r UMAP}
# If you haven't installed UMAP, you can do so via reticulate::py_install(packages = 'umap-learn')
pbmc <- RunUMAP(pbmc, dims = 1:10)

```

```{r DimPlot}

# note that you can set `label = TRUE` or use the LabelClusters function to help label individual clusters

DimPlot(pbmc, reduction = "umap", label = TRUE)

```

You can save the object at this point so that it can easily be loaded back in without having to rerun the computationally intensive steps performed above, or easily shared with collaborators.

saveRDS(pbmc, file = "./output/pbmc_tutorial.rds")

### Finding Differentially Expressed Features (Cluster Biomarkers)

By default, Seurat identifes positive and negative markers of a single cluster (specified in `ident.1`), compared to all other cells. `FindAllMarkers` automates this process for all clusters, but you can also test groups of clusters vs. each other, or against all cells.

The `min.pct` argument requires a feature to be detected at a minimum percentage in either of the two groups of cells, and the `thresh.test` argument requires a feature to be differentially expressed (on average) by some amount between the two groups. You can set both of these to 0, but with a dramatic increase in time - since this will test a large number of features that are unlikely to be highly discriminatory. As another option to speed up these computations, `max.cells.per.ident` can be set. This will downsample each identity class to have no more cells than whatever this is set to. While there is generally going to be a loss in power, the speed increases can be significiant and the most highly differentially expressed features will likely still rise to the top.

```{r Find Cluster 1 Markers}

# find all markers of cluster 1
cluster1.markers <- FindMarkers(pbmc, ident.1 = 1, min.pct = 0.25)
head(cluster1.markers, n = 5)

```

```{r Find Markers Distinguishing cluster 5 from 0 and 3}

# find all markers distinguishing cluster 5 from clusters 0 and 3
cluster5.markers <- FindMarkers(pbmc, ident.1 = 5, ident.2 = c(0, 3), min.pct = 0.25)
head(cluster5.markers, n = 5)

```

```{r Find Markers for Every Cluster compared to All Remaining Cells, Report only the Positive Ones}

# find markers for every cluster compared to all remaining cells, report only the positive ones
pbmc.markers <- FindAllMarkers(pbmc, only.pos = TRUE, min.pct = 0.25, logfc.threshold = 0.25)
pbmc.markers %>% group_by(cluster) %>% top_n(n = 2, wt = avg_logFC)

```

Seurat has several tests for differential expression which can be set with the `test.use` parameter (see our DE vignette for details). For example, the ROC test returns the ‘classification power’ for any individual marker (ranging from 0 - random, to 1 - perfect).

```{r Find Cluster Markers}

cluster1.markers <- FindMarkers(pbmc, ident.1 = 0, logfc.threshold = 0.25, test.use = "roc", only.pos = TRUE)

```

Several tools are included for visualizing marker expression. `VlnPlot` (shows expression probability distributions across clusters), and `FeaturePlot` (visualizes feature expression on a tSNE or PCA plot) are the most commonly used visualizations. `RidgePlot`, `CellScatter`, and `DotPlot` are additional methods to view the dataset.


```{r Visualize Marker Expression - Violin Plot}

VlnPlot(pbmc, features = c("MS4A1", "CD79A"))

```

```{r Plot Raw Counts in a Violin Plot}

# you can plot raw counts as well
VlnPlot(pbmc, features = c("NKG7", "PF4"), slot = "counts", log = TRUE)

```

```{r Feature Plots for Several Genes}

FeaturePlot(pbmc, features = c("MS4A1", "GNLY", "CD3E", "CD14", "FCER1A", "FCGR3A", "LYZ", "PPBP", "CD8A"))

```

`DoHeatmap` generates an expression heatmap for given cells and features. In this case, we are plotting the top 10 markers (or all markers if less than 20) for each cluster.

```{r Plot the Top 10 Markers in a Heatmap}

top10 <- pbmc.markers %>% group_by(cluster) %>% top_n(n = 10, wt = avg_logFC)
DoHeatmap(pbmc, features = top10$gene) + NoLegend()

```

### Assigning Cell Type Identity to Clusters
Fortunately, in the case of this dataset, we can use canonical markers to easily match the unbiased clustering to known cell types:

Cluster ID	Markers	Cell Type
0	IL7R, CCR7	Naive CD4+ T
1	IL7R, S100A4	Memory CD4+
2	CD14, LYZ	CD14+ Mono
3	MS4A1	B
4	CD8A	CD8+ T
5	FCGR3A, MS4A7	FCGR3A+ Mono
6	GNLY, NKG7	NK
7	FCER1A, CST3	DC
8	PPBP	Platelet

```{r Assigning Cell Type ID to Clusters}

new.cluster.ids <- c("Naive CD4 T", "Memory CD4 T", "CD14+ Mono", "B", "CD8 T", "FCGR3A+ Mono", 
    "NK", "DC", "Platelet")
names(new.cluster.ids) <- levels(pbmc)
pbmc <- RenameIdents(pbmc, new.cluster.ids)
DimPlot(pbmc, reduction = "umap", label = TRUE, pt.size = 0.5) + NoLegend()


```

saveRDS(pbmc, file = "./output/pbmc3k_final.rds")
